File: ga/glm4_result_2024_6/result/1968D_cpp_glm/24_1968D.cpp
use std::io;

fn calculate(p: &Vec<usize>, a: &Vec<usize>, pos: usize, k: usize) -> u64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = std::cmp::max(mx, cur + (k as u64) * (a[pos] as u64));
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    mx
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let t: usize = input.trim().parse().expect("Failed to parse test cases");

    for _ in 0..t {
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let mut values = input.trim().split_whitespace().map(|s| s.parse().expect("Failed to parse value"));
        let n: usize = values.next().unwrap();
        let k: usize = values.next().unwrap();
        let ps1: usize = values.next().unwrap();
        let ps2: usize = values.next().unwrap();

        let mut p = Vec::with_capacity(n);
        let mut a = Vec::with_capacity(n);

        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let p_values: Vec<usize> = input.trim().split_whitespace().map(|s| s.parse().expect("Failed to parse p value")).collect();
        p.extend(p_values.iter().map(|&x| x - 1));

        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let a_values: Vec<usize> = input.trim().split_whitespace().map(|s| s.parse().expect("Failed to parse a value")).collect();
        a.extend(a_values.iter().cloned());

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_cpp_mini/24_1968D.rs
use std::io::{self, BufRead};

fn calculate(p: &Vec<usize>, a: &Vec<i64>, mut pos: usize, mut k: i64) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + k * a[pos]);
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    
    mx
}

fn main() {
    let stdin = io::stdin();
    let handle = stdin.lock();
    let reader = handle.lines();
    
    let mut iter = reader.map(|line| line.unwrap()).peekable();
    
    let t: usize = iter.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let line = iter.next().unwrap();
        let mut nums = line.split_whitespace().map(|x| x.parse::<i64>().unwrap());
        let n = nums.next().unwrap() as usize;
        let k = nums.next().unwrap();
        let ps1 = nums.next().unwrap() as usize - 1; // Convert ps1 to 0-based index
        let ps2 = nums.next().unwrap() as usize - 1; // Convert ps2 to 0-based index
        
        let mut p = vec![0; n];
        for i in 0..n {
            let line = iter.next().unwrap();
            p[i] = line.parse::<usize>().unwrap() - 1; // Convert to 0-based index
        }
        
        let mut a = vec![0; n];
        for i in 0..n {
            let line = iter.next().unwrap();
            a[i] = line.parse::<i64>().unwrap();
        }
        
        let ans1 = calculate(&p, &a, ps1, k);
        let ans2 = calculate(&p, &a, ps2, k);
        
        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}